### PR TITLE
Integration with ML4HT and tensorflow 2.5

### PR DESCRIPTION
**Pull request by lucidtronix**

### CODE REVIEW

File: ml4h/models/merge_blocks.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 import tensorflow as tf
 import tensorflow.keras.backend as K
-import tensorflow_probability as tfp
+# import tensorflow_probability as tfp
```
Comment (ndiamant) [inline ml4h/models/merge_blocks.py:5]: do we need to upgrade tensorflow_probability?

File: ml4h/models/legacy_models.py [hunk 2]
```diff
@@ -30,7 +30,7 @@
 from tensorflow.keras.layers import SeparableConv1D, SeparableConv2D, DepthwiseConv2D, Concatenate, Add
 from tensorflow.keras.layers import GlobalAveragePooling1D, GlobalAveragePooling2D, GlobalAveragePooling3D
 #from tensorflow.keras.layers.experimental.preprocessing import RandomRotation, RandomZoom, RandomContrast
-import tensorflow_probability as tfp
+# import tensorflow_probability as tfp
```
Comment (ndiamant) [inline ml4h/models/legacy_models.py:33]: need this for some model blocks i think

File: ml4h/ml4ht_integration/tensor_map.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        )
+        dset = SampleGetterDataset(
+            [1_path_to_hd5.hd5, 2_path_to_hd5.hd5, ...],
+            sg,
+        )
+        data_loader = torch.utils.data.DataLoader(
+            dset, batch_size=128, collate_fn=numpy_collate_fn,
+        )
+
+        model.fit(iter(data_loader))  # train for one epoch
```
Comment (ndiamant) [inline ml4h/ml4ht_integration/tensor_map.py:57]: Is this still true, or do you have to make a tensorflow dataset from the data loader?

File: ml4h/defines.py [hunk 4]
```diff
@@ -161,3 +162,11 @@ def dataset_name_from_meaning(group: str, fields: List[str]) -> str:
     if group is None:
         return joined
     return group + HD5_GROUP_CHAR + joined
+
+
+class TensorGeneratorABC(ABC):
+    @abstractmethod
+    def __iter__(self) -> Iterator:
+        pass
```
Comment (ndiamant) [inline ml4h/defines.py:170]: I think this might be equivalent to something a little cleaner
```
from collections.abc import Iterable
TensorGeneratorABC =  Iterator
```

File: docker/vm_boot_images/config/tensorflow-requirements.txt [hunk 5]
```diff
@@ -16,8 +14,10 @@ bokeh
 Pillow==8.1.1
 notebook
 pytest
-tensorflow-addons==0.9.1
-tensorflow_probability==0.9.0
+tensorflow==2.5.0
```
Comment (ndiamant) [inline docker/vm_boot_images/config/tensorflow-requirements.txt:17]: woo!

**Review by ndiamant — CHANGES_REQUESTED** [review]
A few small changes and questions. Exciting!!!

File: ml4h/models/legacy_models.py (hunk 2 above)
Comment (lucidtronix) [reply ml4h/models/legacy_models.py]: yes should still work will put it back

File: ml4h/ml4ht_integration/tensor_map.py (hunk 3 above)
Comment (lucidtronix) [reply ml4h/ml4ht_integration/tensor_map.py:57]: Thats what I did now, but if num_workers is too high I get Queue.empty errors in dataloader

File: docker/vm_boot_images/config/tensorflow-requirements.txt (hunk 5 above)
Comment (lucidtronix) [reply docker/vm_boot_images/config/tensorflow-requirements.txt:17]: :)

File: tests/test_models.py [hunk 6]
```diff
@@ -472,21 +482,19 @@ def test_brain_seg(self, tmpdir):
...
-                output_folder=str(tmpdir), run_id='brain_seg_test',
-                inspect_model=True, inspect_show_labels=True,
-            )
-            test_data, test_labels, test_paths = big_batch_from_minibatch_generator(
-                generate_test, 12,
-            )
-        finally:
-            generate_train.kill_workers()
-            generate_test.kill_workers()
-            generate_valid.kill_workers()
```
Comment (ndiamant) [inline tests/test_models.py:489]: nice to lose this :)

**Review by ndiamant — APPROVED** [review]
approved after small style fixes

**Comment by paolodi** [comment]
Don't think I can find enough emoticons to cheer for this ❤️ 💌 💓 🥳 . Also TF 2.5 is needed to compile the contrastive loss that we are using on echos https://keras.io/examples/vision/supervised-contrastive-learning/!

File: tests/test_models.py (hunk 6 above)
Comment (lucidtronix) [reply tests/test_models.py:489]: :)

File: ml4h/recipes.py [hunk 7]
```diff
@@ -494,6 +498,57 @@ def train_paired_model(args):
...
+    args.num_workers = 0
+    tsv_style_is_genetics = 'genetics' in args.tsv_style
+    sample_set = None
+    if args.sample_csv is not None:
+        with open(args.sample_csv, 'r') as csv_file:
+            sample_ids = [row[0] for row in csv.reader(csv_file)]
+            sample_set = set(sample_ids[1:])
+    _, encoders, _, _ = block_make_multimodal_multitask_model(**args.__dict__)
+    latent_dimensions = args.dense_layers[-1]
+    for e in encoders:
```
[the remark this answers is not in the document]
Comment (lucidtronix) [reply ml4h/recipes.py:511]: Done!

File: ml4h/recipes.py [hunk 8]
```diff
@@ -494,6 +498,57 @@ def train_paired_model(args):
...
+        stats = Counter()
+        inference_tsv = _hidden_file_name(args.output_folder, e.name, args.id, '.tsv')
+        logging.info(f'Will write encodings from {e.name} to: {inference_tsv}')
+        # hard code batch size to 1 so we can iterate over file names and generated tensors together in the tensor_paths for loop
+        tensor_paths = [
+            os.path.join(args.tensors, tp) for tp in sorted(os.listdir(args.tensors))
+            if os.path.splitext(tp)[-1].lower() == TENSOR_EXT and (sample_set is None or os.path.splitext(tp)[0] in sample_set)
```
[the remark this answers is not in the document]
Comment (lucidtronix) [reply ml4h/recipes.py:518]: Done!

File: benchmarks/benchmark.py [hunk 9]
```diff
@@ -64,8 +64,8 @@ def __call__(self, batch_size: int, num_workers: int) -> Generator:
             cache_size=0, paths=self.paths,
         )
         yield gen
-        gen.kill_workers()
-        del gen
+        #gen.kill_workers()
+        #del gen
```
[the remark this answers is not in the document]
Comment (lucidtronix) [reply benchmarks/benchmark.py:68]: Done!

File: ml4h/defines.py (hunk 4 above)
Comment (lucidtronix) [reply ml4h/defines.py:189]: Done

**Comment by curlup** [comment]
notebooks and setup py still ask for `ml4ht`
